### ISSUE TITLE
Replace type switches for tx and batch for reflection, it's a maintenance nightmare at the moment.

### ISSUE THREAD

**Issue by ruseinov** [iov-one/weave#212]

**Comment by ethanfrey** [comment]
@ruseinov Can you please explain a bit of what you suggest?

There is currently a manual switch in eg. https://github.com/iov-one/weave/blob/master/cmd/bcpd/app/tx.go#L34-L67 that must be updated for every message added to the protobuf file. We need that to take the sum type (protobuf-generated) and extract the embedded message that it refers to (our defined type).

I assume the desired replacement would be:

```golang
func (tx *Tx) GetMsg() (weave.Msg, error) {
  sum := tx.GetSum()
  msg, err := ExtractMsgFromSum(sum)
  return msg, err
}
```

And we need to write one `ExtractMsgFromSum` function that can be used in all applications. Is that the desired solution?

**Comment by ethanfrey** [comment]
I guess we can check that `sum` is a struct with exactly one field, and return an error otherwise. 
If successful, then extract the value of that one field and try to cast it to Msg, returning a proper error if anything doesn't work along the way.

https://golang.org/pkg/reflect/#Type
https://golang.org/pkg/reflect/#Value.Field

**Comment by ethanfrey** [comment]
Auto-generating from protobuf is also valid solution.

Right now, we need to update protobuf, compile it, and update two unwrap methods (one for Tx, one for Batch). That is annoying and error-prone busy work.

**Comment by ruseinov** [comment]
@ethanfrey I'd be happy to look into that, been thinking of reflection as we just need to check for an interface of a certain kind. Generating this from protobuf is going to involve a template, and while this is a valid solution it's so much harder to maintain as templates are not parsed or autocompleted in the ide.

**Comment by ruseinov** [comment]
@ethanfrey Here's an example of such interface https://github.com/iov-one/weave/blob/master/cmd/bcpd/app/codec.pb.go#L76

Somehow missed your comment a while ago

**Comment by ethanfrey** [comment]
The protobuf autogenerating solution came up in backlog grooming. Not my favorite but we decided it was also a solution, so I added it to leave flexibility to whoever implements it

**Comment by ethanfrey** [comment]
Closed by #313
